### PR TITLE
[FSDP] summon_full_params allows multiple modules

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102692
* #102680
* #102679
* #102672
* #102669
* #102658
* #102657
* __->__ #102655
* #102639
* #102637
* #102508

Since this is a staticmethod, slight enhancement to allow summoning
full params for multiple modules. This is a small UX improvement for things
like testing use cases when we want to run this for multiple FSDP modules and
compare things.

Note that for now, the configuration under which they are summoned is the same.

Differential Revision: [D46334991](https://our.internmc.facebook.com/intern/diff/D46334991/)

### CODE REVIEW

**Comment by rohan-varma** [comment]
Because some ``nn.Module`` are iterable (nn.Sequential), will gate this to be a tuple explicitly for now.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -859,10 +861,23 @@ def summon_full_params(
...
+                    )
+                    for m in module_iter
+                ]
+                for mgr in unshard_contexts:
+                    stack.enter_context(mgr)
+                yield
+        except TypeError:
+            with _unshard_params(
+                module, recurse, writeback, rank0_only, offload_to_cpu, with_grads
+            ):
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:879]: @fegin yeah that makes sense. Since some modules are iterable, I'm planning to more explicitly separate the 2 codepaths by enforcing a `tuple` of modules has to be passed in if we want to do multiple summon.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -834,6 +834,8 @@ def summon_full_params(
             ``rank0_only=True``.
 
         Args:
+            module: Union[nn.Module, Tuple[nn.Module]]: Module or tuple
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:837]: We should use `Tuple[nn.Module, ...]` to denote variable-length tuple of homogeneous type `nn.Module`.

File: test/distributed/fsdp/test_fsdp_unshard_params.py [hunk 3]
```diff
@@ -228,6 +228,26 @@ def test_unshard_params_param_data(self):
             self._test_unshard_params_param_data,
         )
 
+    @skip_if_lt_x_gpu(2)
+    def test_unshard_params_multiple_modules(self):
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_unshard_params.py:232]: Computing the expected numel as the product of the sharded parameter size and the world size is only true in this no padding case, which could be brittle (e.g. if the sharding changes). Should we consider comparing the result of nested calls to `summon_full_params()` each on a single module vs. a single call to `summon_full_params()` on multiple modules at once? (This is what I interpret as the direct semantics of generalizing the API.)

I am also curious if this is testing the expected use case for adding this feature. When does a user have multiple distinct FSDP modules? I was thinking about the case where we want to use `summon_full_params(recurse=False)` on multiple modules belonging to the same module tree.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -793,7 +793,7 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
     @staticmethod
     @contextlib.contextmanager
     def summon_full_params(
-        module: nn.Module,
+        module: Union[nn.Module, Tuple[nn.Module]],
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:796]:
```suggestion
        module: Union[nn.Module, Tuple[nn.Module, ...]],
```

File: test/distributed/fsdp/test_fsdp_unshard_params.py (hunk 3 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_unshard_params.py:232]: > like testing use cases when we want to run this for multiple FSDP modules and
compare things.

Does this imply that this feature is meant for testing?

Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_unshard_params.py:232]: @awgu I was also thinking of the use case of 2 separate encoders both wrapped with FSDP, and users want to call summon_full_params on both of them.
